### PR TITLE
lib: nrf_modem: trace: fixes for uart trace backend++

### PR DESCRIPTION
**Pull request by eivindj-nordic**
If the UART trace backend has flow control enabled, the trace backend might abort the current send, hence 0 bytes will be written. Return an error in this case to not assert on the trace backend returning 0 bytes.

### CODE REVIEW

**Comment by lemrey** [comment]
Isn't that an issue with how we define the `trace_backend_write` interface?
Why do we allow that to return `0`? Is it necessary to differentiate between `0` and errors?

**Comment by eivindj-nordic** [comment]
> Isn't that an issue with how we define the `trace_backend_write` interface? Why do we allow that to return `0`? Is it necessary to differentiate between `0` and errors?

It depends. The alternative is to let the trace backend loop forever in this case instead, and log the warning there. But this will happen in some cases, where we do not want to assert.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -381,7 +381,11 @@ Modem libraries
 
 * :ref:`nrf_modem_lib_readme`:
 
-  * Updated the :c:func:`nrf_modem_lib_shutdown` function to allow the modem to be in flight mode (``CFUN=4``) when shutting down the modem.
+  Updated:
+
+    * The :c:func:`nrf_modem_lib_shutdown` function to allow the modem to be in flight mode (``CFUN=4``) when shutting down the modem.
+    * The trace backends can now return ``-EAGAIN`` if the write operation can be retried.
+    * The trace backend deinit function is expected to succeed and can no longer return an error.
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:388]: There is no function called `trace_backend_deinit`, but there is a `deinit` field in the `nrf_modem_lib_trace_backend` struct with a template for the function signature. I can add `deinit` in backticks.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 2]
```diff
@@ -381,7 +381,11 @@ Modem libraries
 
 * :ref:`nrf_modem_lib_readme`:
 
-  * Updated the :c:func:`nrf_modem_lib_shutdown` function to allow the modem to be in flight mode (``CFUN=4``) when shutting down the modem.
+  Updated:
+
+    * The :c:func:`nrf_modem_lib_shutdown` function to allow the modem to be in flight mode (``CFUN=4``) when shutting down the modem.
+    * The trace backends can now return ``-EAGAIN`` if the write operation can be retried.
+    * The trace backends ``deinit`` function is expected to succeed and can no longer return an error.
```
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:388]: @eivindj-nordic : There is one more entry : _Added CEREG event tracking to lte_connectivity_ that should be part of :ref:`nrf_modem_lib_readme`: . Could you move this entry here.

Thanks for fixing this.

Comment (eivindj-nordic) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:390]: done

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 3]
```diff
@@ -381,7 +381,13 @@ Modem libraries
...
 * :ref:`nrf_modem_lib_readme`:
 
-  * Updated the :c:func:`nrf_modem_lib_shutdown` function to allow the modem to be in flight mode (``CFUN=4``) when shutting down the modem.
+  * Added CEREG event tracking to ``lte_connectivity``.
+
+  Updated:
+
+    * The :c:func:`nrf_modem_lib_shutdown` function to allow the modem to be in flight mode (``CFUN=4``) when shutting down the modem.
+    * The trace backends can now return ``-EAGAIN`` if the write operation can be retried.
+    * The trace backends ``deinit`` function is expected to succeed and can no longer return an error.
```
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:390]:
```suggestion
  * Added CEREG event tracking to ``lte_connectivity``.

  * Updated:

    * The :c:func:`nrf_modem_lib_shutdown` function to allow the modem to be in flight mode (``CFUN=4``) when shutting down the modem.
    * The trace backends can now return ``-EAGAIN`` if the write operation can be retried.
    * The trace backends ``deinit`` function is expected to succeed and can no longer return an error.
```
nit

**Review by divipillai — APPROVED** [review]

**Review by nordic-mare — APPROVED** [review]

File: include/modem/trace_backend.h [hunk 4]
```diff
@@ -40,11 +40,8 @@ struct nrf_modem_lib_trace_backend {
 
 	/**
 	 * @brief Deinitialize the compile-time selected trace backend.
-	 *
-	 * @return 0 If the operation was successful.
-	 *         Otherwise, a (negative) error code is returned.
 	 */
-	int (*deinit)(void);
+	void (*deinit)(void);
```
Comment (lemrey) [inline include/modem/trace_backend.h:44]: I think it's best to keep it returning an `int`, even if current implementations can't fail. That is still a possibility if they change or new ones are added; by removing the return value and the log it's impossible to tell if this operation failed.

**Review by lemrey — COMMENTED** [review]
One small comment otherwise looks good

File: include/modem/trace_backend.h (hunk 4 above)
Comment (eivindj-nordic) [reply include/modem/trace_backend.h]: updated

**Review by lemrey — APPROVED** [review]

**Pull request merged**